src/physics/adv_fluxcorr.f90

```diff
@@ -0,0 +1,193 @@
+        end where
+        
+        where (w(:,kms:kme-1,:) > 0)
+            qmax(:,kms+1:kme,:) = max(q(its-1:ite+1,kms:kme-1,jts-1:jte+1),qmax(:,kms+1:kme,:))
+            qmin(:,kms+1:kme,:) = min(q(its-1:ite+1,kms:kme-1,jts-1:jte+1),qmin(:,kms+1:kme,:))
+        else where (w(:,kms:kme-1,:) < 0)
+            qmax(:,kms:kme-1,:) = max(q(its-1:ite+1,kms+1:kme,jts-1:jte+1),qmax(:,kms:kme-1,:))
+            qmin(:,kms:kme-1,:) = min(q(its-1:ite+1,kms+1:kme,jts-1:jte+1),qmin(:,kms:kme-1,:))
+        end where
+
+        
+        where(flux_in  > (qmax-temp))  scale_in = max(0.,(qmax-temp)/(flux_in + 1.E-15))
+        where(flux_out > (temp-qmin)) scale_out = max(0.,(temp-qmin)/(flux_out+ 1.E-15))
+              
+        ! We are now done with merged fluxes, so they can be overwritten to save memory
+        ! flux now represents the normalized fluxes
+        where(flux_x(its:ite+1,:,:) > 0)
+            flux_x(its:ite+1,:,:) = min(scale_in(its:ite+1,:,:),scale_out(its-1:ite,:,:))*flux_x(its:ite+1,:,:)
+        else where(flux_x(its:ite+1,:,:) < 0)
+            flux_x(its:ite+1,:,:) = min(scale_out(its:ite+1,:,:),scale_in(its-1:ite,:,:))*flux_x(its:ite+1,:,:)
+        end where
+
+        where(flux_y(:,:,jts:jte+1) > 0)
+            flux_y(:,:,jts:jte+1) = min(scale_in(:,:,jts:jte+1),scale_out(:,:,jts-1:jte))*flux_y(:,:,jts:jte+1)
+        else where(flux_y(:,:,jts:jte+1) < 0)
+            flux_y(:,:,jts:jte+1) = min(scale_out(:,:,jts:jte+1),scale_in(:,:,jts-1:jte))*flux_y(:,:,jts:jte+1)
+        end where
+        
+        where(flux_z(:,kms+1:kme,:) > 0)
+            flux_z(:,kms+1:kme,:) = min(scale_in(:,kms+1:kme,:),scale_out(:,kms:kme-1,:))*flux_z(:,kms+1:kme,:)
+        else where(flux_z(:,kms+1:kme,:) < 0)
+            flux_z(:,kms+1:kme,:) = min(scale_out(:,kms+1:kme,:),scale_in(:,kms:kme-1,:))*flux_z(:,kms+1:kme,:)
+        end where
+
+        !Finally, compute the output fluxes
+        flux_x = flux_x + upwind_flux_x
+        flux_y = flux_y + upwind_flux_y
+        flux_z = flux_z + upwind_flux_z
+        
+    end subroutine WRF_flux_corr
+
+    subroutine upwind_flux3(q,u,v,w,flux_x,flux_z,flux_y,jaco,dz,rho)
+        implicit none
+        real, dimension(ims:ime,  kms:kme,jms:jme),    intent(in)      :: q
+        real, dimension(its-1:ite+2,  kms:kme,jts-1:jte+1),  intent(in)    :: u
+        real, dimension(its-1:ite+1,  kms:kme,jts-1:jte+2),  intent(in)    :: v
+        real, dimension(its-1:ite+1,  kms:kme,jts-1:jte+1),  intent(in) :: w, jaco, dz, rho
+
+        real, dimension(its-1:ite+2,kms:kme,jts-1:jte+1),intent(inout)          :: flux_x
+        real, dimension(its-1:ite+1,kms:kme,jts-1:jte+2),intent(inout)          :: flux_y
+        real, dimension(its-1:ite+1,kms:kme+1,jts-1:jte+1),intent(inout)    :: flux_z
+        
+        real, dimension(ims:ime,  kms:kme,jms:jme) :: dumb_q
+
+        !When using RK3, we may have a time step derived using a CFL constraint larger than 1
```

Comment:
would it make sense to have an `if CFL>1` then do it twice? otherwise just do it once?

Reply:
Yes, this would be the most logical from an algorithm perspective. My thinking was that the advection code outside of the driver is more or less insulated from the time options at the moment, so you don't have access to that information if the CFL is greater than 1. Plus, the computational load of calculating upwind fluxes twice is pretty low, and running this flux correction scheme without RK3 stepping may be unnecessary.